### PR TITLE
Bugfix various docker startup issues

### PR DESCRIPTION
**Pull request by lunkwill42**
We are observing reams of various issues with getting the containers up and running for running the integration tests, even when following the guidelines from SUNET's wiki. 

The level of the problems differs when using Linux and when using OSX as the host platforms, but most of them are concerned with missing error handling in shell scripts, missing file permissions and outdated configuration of the PostgreSQL containers.

It seems we are able to get most things running with these changes, so maybe we can actually get to running the test suite soon.

### CODE REVIEW

File: docker/api/Dockerfile [hunk 1]
```diff
@@ -21,6 +21,7 @@ RUN /opt/cnaas/cnaas-setup.sh $BUILDBRANCH
 COPY nosetests.sh /opt/cnaas/
 COPY exec-pre-app.sh /opt/cnaas/
 COPY config/uwsgi.ini /opt/cnaas/venv/cnaas-nms/
+RUN chown www-data /opt/cnaas/venv/cnaas-nms/uwsgi.ini
```
Comment (indy-independence) [inline docker/api/Dockerfile:24]: chown root:www-data?
chmod 640

File: docker/api/Dockerfile [hunk 2]
```diff
@@ -34,17 +35,18 @@ COPY client.html /opt/cnaas/static
 RUN unlink /etc/nginx/sites-enabled/default
 RUN ln -s /etc/nginx/sites-available/nginx_app.conf /etc/nginx/sites-enabled/default
 RUN chown www-data:www-data /var/log/nginx
-RUN chown www-data:www-data /etc/cnaas-nms/repository.yml
+RUN chown www-data:www-data /etc/cnaas-nms/*.yml
```
Comment (indy-independence) [inline docker/api/Dockerfile:38]: chown root:www-data?
chmod 640

File: docker/api/Dockerfile [hunk 3]
```diff
@@ -34,17 +35,18 @@ COPY client.html /opt/cnaas/static
...
 RUN chown -R  www-data:www-data /var/log/nginx/
 RUN chown -R www-data:www-data /var/lib/nginx
 RUN chown www-data:www-data /var/lib/nginx/
+RUN chown -R www-data:www-data /etc/nginx/
```
Comment (indy-independence) [inline docker/api/Dockerfile:42]: chown root:www-data?
chmod -R u=rwX,g=rX,o=

File: docker/api/Dockerfile [hunk 4]
```diff
@@ -34,17 +35,18 @@ COPY client.html /opt/cnaas/static
...
 # Give permission for API to clone/sync repos
-RUN chown www-data:www-data /opt/cnaas
+RUN chown www-data:www-data -R /opt/cnaas
```
Comment (indy-independence) [inline docker/api/Dockerfile:44]: chown root:www-data
chmod -R u=rwX,g=rX,o=
and remove the two following lines?

**Comment by katsel** [comment]
I setup and enabled Virtualbox switches as according to the wiki, and I can confirm that the integration tests are **running** on my CentOS 8 system with these changes.

File: docker/api/Dockerfile (hunk 2 above)
Comment (katsel) [reply docker/api/Dockerfile:38]: Changed to
```
RUN chown root:www-data /etc/cnaas-nms/*.yml && \
    chmod 640 /etc/cnaas-nms/*.yml
```

I get error 500 on the integration tests.

Docker log:
```
cnaas_api_1       | 2020-11-18 09:00:13,837 DEBG 'uwsgi' stdout output:
cnaas_api_1       | pcre jit disabled
cnaas_api_1       | detected number of CPU cores: 8
cnaas_api_1       | current working directory: /
cnaas_api_1       | detected binary path: /usr/local/bin/uwsgi
cnaas_api_1       | chdir() to /opt/cnaas/venv/cnaas-nms/src/
cnaas_api_1       | your memory page size is 4096 bytes
cnaas_api_1       | detected max file descriptor number: 1048576
cnaas_api_1       | - async cores set to 1000 - fd table size: 1048576
cnaas_api_1       | lock engine: pthread robust mutexes
cnaas_api_1       | thunder lock: disabled (you can enable it with --thunder-lock)
cnaas_api_1       | uwsgi socket 0 bound to UNIX address /tmp/uwsgi.sock fd 6
cnaas_api_1       | Python version: 3.7.3 (default, Jul 25 2020, 13:03:44)  [GCC 8.3.0]
cnaas_api_1       | PEP 405 virtualenv detected: /opt/cnaas/venv
cnaas_api_1       | Set PythonHome to /opt/cnaas/venv
cnaas_api_1       | Python main interpreter initialized at 0x55d5aefe7930
cnaas_api_1       | python threads support enabled
cnaas_api_1       | your server socket listen backlog is limited to 100 connections
cnaas_api_1       | your mercy for graceful operations on workers is 60 seconds
cnaas_api_1       | mapped 21036928 bytes (20543 KB) for 1000 cores
cnaas_api_1       | *** Operational MODE: async ***
cnaas_api_1       | *** uWSGI is running in multiple interpreter mode ***
cnaas_api_1       | spawned uWSGI master process (pid: 10)
cnaas_api_1       | spawned uWSGI worker 1 (pid: 11, cores: 1000)
cnaas_api_1       | spawned uWSGI mule 1 (pid: 12)
cnaas_api_1       | 
cnaas_api_1       | 2020-11-18 09:00:13,841 DEBG fd 9 closed, stopped monitoring <POutputDispatcher at 139666061079544 for <Subprocess at 139666061398384 with name nginx in state STARTING> (stderr)>
cnaas_api_1       | 2020-11-18 09:00:13,854 DEBG 'uwsgi' stdout output:
cnaas_api_1       | running "/opt/cnaas/exec-pre-app.sh" (pre app)...
cnaas_api_1       | 
cnaas_api_1       | 2020-11-18 09:00:13,859 DEBG 'uwsgi' stdout output:
cnaas_api_1       | /opt/cnaas/exec-pre-app.sh: 15: /opt/cnaas/exec-pre-app.sh: cannot create /etc/cnaas-nms/repository.yml: Permission denied
```

Comment (indy-independence) [reply docker/api/Dockerfile:38]: Ah sorry I forgot about that, but I think it's fine if we change so www-data:www-data is owner of /etc/cnaas-nms/

**Comment by katsel** [comment]
Changes:

- automate Docker volume creation for integration tests
- automate installation of the JWT cert for integration tests
- update file permissions on the Docker API files that caused issues for @lunkwill42 and me
- reduce number of layers on the Dockerfile where appropriate
- minor code formatting changes

Changes in file permissions are to the best of my knowledge and following @indy-independence's remarks.
I confirm that with these changes, integration tests are running (HTTP status 200) and there are no major errors on the Docker log.

Observations:
- a teardown function would be useful: allow deleting volumes and Docker instances after successful integration test run

**Comment by indy-independence** [comment]
Looks good!

**Pull request merged**